### DISCUSSION TITLE
How to load Style by reflection

### DISCUSSION THREAD

**Discussion by Maklith** [AvaloniaUI/Avalonia#15056] [Q&A] [answered]
I want to load a Style by reflection, but I'm not sure how to get the ServiceProvider I need
```cs
if (type.FullName=="CompiledAvaloniaXaml.!AvaloniaResources")
            {
                foreach (var methodInfo in type.GetMethods())
                {
                   if(methodInfo.ReturnType==typeof(Styles))
                   {
                       object serviceProvider = null;//this
                       var style = methodInfo.Invoke(null, new object?[] { serviceProvider });
                       Application.Current.Styles.Add((IStyle)style);
                   }
                }
            }
```
need a sp like this
![6c56d90618a8bdc7ab309d185054ff30](https://github.com/AvaloniaUI/Avalonia/assets/42534870/73410950-95fa-414d-aa8a-8053ca7e673b)

**Comment by Maklith** [comment]
Answer
```cs
if (type.FullName=="CompiledAvaloniaXaml.!AvaloniaResources")
            {
                foreach (var methodInfo in type.GetMethods())
                {
                   if(methodInfo.Name.StartsWith("Populate:"))
                   {
                       object rootServiceProviderV2 = XamlIlRuntimeHelpers.CreateRootServiceProviderV2();
                       var style = methodInfo.Invoke(null, new object?[] { rootServiceProviderV2,Application.Current.Styles});
                   }
                }
            }
```

**Reply by Maklith** [reply]
This solution applies only to single-tier resources

**Comment by maxkatz6** [comment]
Why do you need reflection here?
You already can load styles with StyleInclude by their Uri, that can be dynamically generated.
Like, `new Uri("avares://Assembly/Style.axaml")`. Same even works with Fluent one - `new Uri("avares://Avalonia.Themes.Fluent/FluentTheme.xaml")` (as an alternative to `<FluentTheme/>`).

**Reply by maxkatz6** [reply]
Where StyleInclude would still use reflection under the hood, but at least in a more controlled way.

**Reply by maxkatz6** [reply]
As `CompiledAvaloniaXaml.!AvaloniaResources` IS NOT a public API, and might be broken in minor releases.

**Reply by Maklith** [reply]
> As `CompiledAvaloniaXaml.!AvaloniaResources` IS NOT a public API, and might be broken in minor releases. 

But what I need is a way to load the style in assemly, because I want to use AssemblyLoadContext which contains Avalonia content。
Not public API is also ok because I am trying (PR #13974)

**Reply by Maklith** [reply, marked as answer]
I'm sorry for the problem, because the tests found that loading a Style in a dynamic assembly can be done with just use the Url
```cs
        var styles = new Styles();
        var styleInclude = new StyleInclude(new Uri("avares://UnloadableAssemblyLoadContextPlug", UriKind.Absolute));
        styleInclude.Source=new Uri("ControlStyle.axaml", UriKind.Relative);
        styles.Add(styleInclude);
        Application.Current.Styles.Add(styles);
```

**Comment by Maklith** [comment]
> Why do you need reflection here? You already can load styles with StyleInclude by their Uri, that can be dynamically generated. Like, `new Uri("avares://Assembly/Style.axaml")`. Same even works with Fluent one - `new Uri("avares://Avalonia.Themes.Fluent/FluentTheme.xaml")` (as an alternative to `<FluentTheme/>`).

If I use AssemblyLoadContext to load an Assemly containing Style, it seems that I only use `AvaloniaXamlLoader.Load(new) Uri("avares://UnloadableAssemblyLoadContextPlug/TestControl.axaml"));` Is invalid, no style was loaded

**Reply by jp2masa** [reply]
For extensions, I think the easiest way to load styles is the following:

1. Generate a class for the styles (e.g. [FluentTheme.xaml](https://github.com/AvaloniaUI/Avalonia/blob/bad2e8dbdcd3bd4ffd75a35fe94c20ff8ca2b880/src/Avalonia.Themes.Fluent/FluentTheme.xaml) and [FluentTheme.xaml.cs](https://github.com/AvaloniaUI/Avalonia/blob/bad2e8dbdcd3bd4ffd75a35fe94c20ff8ca2b880/src/Avalonia.Themes.Fluent/FluentTheme.xaml.cs))
2. Create an interface like this:
    ```cs
    public interface IStyleProvider
    {
        IStyle CreateStyle();
    }
    ```
3. An example implementation for `FluentTheme` would look like this:
    ```cs
    public interface FluentThemeStyleProvider : IStyleProvider
    {
        public IStyle CreateStyle() =>
            new FluentTheme();
    }
    ```

**Reply by Maklith** [reply]
> For extensions, I think the easiest way to load styles is the following:
> 
> 1. Generate a class for the styles (e.g. [FluentTheme.xaml](https://github.com/AvaloniaUI/Avalonia/blob/bad2e8dbdcd3bd4ffd75a35fe94c20ff8ca2b880/src/Avalonia.Themes.Fluent/FluentTheme.xaml) and [FluentTheme.xaml.cs](https://github.com/AvaloniaUI/Avalonia/blob/bad2e8dbdcd3bd4ffd75a35fe94c20ff8ca2b880/src/Avalonia.Themes.Fluent/FluentTheme.xaml.cs))
> 2. Create an interface like this:
>     ```cs
>     public interface IStyleProvider
>     {
>         IStyle CreateStyle();
>     }
>     ```
> 3. An example implementation for `FluentTheme` would look like this:
>     ```cs
>     public interface FluentThemeStyleProvider : IStyleProvider
>     {
>         public IStyle CreateStyle() =>
>             new FluentTheme();
>     }
>     ```
This may seem like a good solution, but is ineffective for styles and resource dictionaries without classes

**Reply by maxkatz6** [reply]
You still can use StyleInclude/ResourceInclude inside of IStyleProvider impl though.
